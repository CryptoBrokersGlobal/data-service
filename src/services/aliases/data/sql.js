const pg = require('knex')({ client: 'pg' });

const LIMIT = 1000;

const selectFromSet = aliasSet =>
  pg.select('alias', 'address', 'duplicates').from({
    counted_aliases: pg({ t: 'txs_10' })
      .select({ alias: 't.alias' })
      .min({ address: 'addr.address' }) // first sender
      .min({ time_stamp: 'txs.time_stamp' }) // first tx timestamp
      .count({ duplicates: 't.sender_uid' }) // count senders grouped by alias
      .leftJoin({ txs: 'txs' }, 'txs.uid', 't.tx_uid')
      .leftJoin({ addr: 'addresses' }, 'addr.uid', 't.sender_uid')
      .whereIn('t.alias', aliasSet)
      .groupBy('t.alias'),
  });

module.exports = {
  get: alias =>
    selectFromSet([alias])
      .clone()
      .toString(),
  mget: aliases =>
    selectFromSet(aliases)
      .clone()
      .toString(),
  search: ({ address, showBroken }) => {
    const q = selectFromSet(
      pg('txs_10')
        .select('alias')
        .where('sender_uid', function() {
          this.select('uid')
            .from('addresses')
            .where('address', address)
            .limit(1);
        })
    )
      .clone()
      .orderBy('time_stamp', 'asc')
      .limit(LIMIT);

    // aliases are considered broken if 'duplicates' not equal to 1
    if (showBroken) {
      return q.toString();
    } else {
      return q
        .clone()
        .where('duplicates', 1)
        .toString();
    }
  },
};
