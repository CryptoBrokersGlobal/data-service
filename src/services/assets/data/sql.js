const pg = require('knex')({ client: 'pg' });

module.exports = ids =>
  pg('assets')
    .select([
      'asset_id',
      'asset_name',
      'description',
      'sender',
      'issue_height',
      'total_quantity',
      'decimals',
      'reissuable',
      'ticker',
      'issue_timestamp'
    ])
    .whereIn('asset_id', ids)
    .toString();
