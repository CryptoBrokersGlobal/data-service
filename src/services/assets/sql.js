const pg = require('knex')({ client: 'pg' });
const { compose } = require('ramda');

const columns = [
  'asset_id',
  'asset_name',
  'description',
  'sender',
  'issue_height',
  'total_quantity',
  'decimals',
  'reissuable',
  'ticker',
  'issue_timestamp',
  'has_script',
];

const mget = ids =>
  pg('assets')
    .select(columns)
    .whereIn('asset_id', ids)
    .toString();

module.exports = {
  get: id => mget([id]),
  mget,
  search: ({ ticker }) => {
    const filter = q => {
      if (ticker === '*') return q.whereNotNull('ticker');
      else return q.where('ticker', ticker);
    };

    return compose(
      q => q.toString(),
      filter
    )(pg('assets').select(columns));
  },
};
