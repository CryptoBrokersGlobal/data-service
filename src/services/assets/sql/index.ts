import * as knex from 'knex';
import { compose } from 'ramda';
import { columns } from './common';
import { searchAssets } from './searchAssets';

const pg = knex({ client: 'pg' });

const getAssetIndex = (asset_id: string) =>
  pg('assets_cte')
    .column('rn')
    .where('asset_id', asset_id);

export const mget = (ids: number[]): string =>
  pg('assets')
    .select(columns)
    .whereIn('asset_id', ids)
    .toString();

export const get = (id: number): string => mget([id]);

export const search = ({
  ticker,
  search,
  after,
  limit,
}: {
  ticker: string;
  search: string;
  after: string;
  limit: number;
}): string => {
  const filter = (q: knex.QueryBuilder) => {
    if (ticker === '*') return q.whereNotNull('ticker');
    else return q.where('ticker', ticker);
  };

  if (typeof ticker !== 'undefined') {
    return compose(
      (q: knex.QueryBuilder) => q.toString(),
      filter
    )(pg('assets').select(columns));
  } else {
    return compose(
      (q: knex.QueryBuilder) => q.toString(),
      (q: knex.QueryBuilder) => (limit ? q.clone().limit(limit) : q),
      (q: knex.QueryBuilder) =>
        after ? q.clone().where('rn', '>', getAssetIndex(after)) : q
    )(searchAssets(search));
  }
};
