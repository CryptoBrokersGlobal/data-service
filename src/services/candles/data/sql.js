const pg = require('knex')({ client: 'pg' });

module.exports = filters =>
  pg('candles')
    .select('*')
    .whereBetween('timestamp', [filters.timeStart, filters.timeEnd]);
