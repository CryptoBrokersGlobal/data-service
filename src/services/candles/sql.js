const {
  compose,
  find,
  findLast,
  map,
  multiply,
  prop,
  sortBy,
} = require('ramda');
const { interval, Unit } = require('../../types');
const {
  div,
  fromMillisecs: intervalFromMillisecs,
} = require('../../utils/interval');
const { add, trunc } = require('../../utils/date');
const pg = require('knex')({ client: 'pg' });

const fields = [
  'time_start',
  'amount_asset_id',
  'price_asset_id',
  'low',
  'high',
  'volume',
  'quote_volume',
  'max_height',
  'txs_count',
  'weighted_average_price',
  'open',
  'close',
  'interval_in_secs',
];

const fieldsWithDecimals = [
  ...fields,
  { a_dec: 'a_dec.decimals' },
  { p_dec: 'p_dec.decimals' },
];

/** highestDividerLessThen :: Interval i => i -> String[] -> i */
const highestDividerLessThen = (inter, dividers) =>
  compose(
    findLast(i => div(inter, i) >= 1),
    sortBy(prop('length')),
    map(d => interval(d).getOrElse(null))
  )(dividers);

/** selectCandles :: { amountAsset: String, priceAsset: String, timeStart: Date, timeEnd: Date, interval: String } -> knex.QueryBuilder */
const selectCandles = ({
  amountAsset,
  priceAsset,
  timeStart,
  timeEnd,
  interval: inter,
}) =>
  pg('candles')
    .select(fields)
    .where('amount_asset_id', amountAsset)
    .where('price_asset_id', priceAsset)
    .where('time_start', '>=', timeStart)
    .where('time_start', '<=', timeEnd)
    .where(
      'interval_in_secs',
      highestDividerLessThen(interval(inter).getOrElse(null), [
        '1m',
        '5m',
        '15m',
        '30m',
        '1h',
        '3h',
        '6h',
        '12h',
        '1d',
      ]).length / 1000
    );

/** periodToQueries :: { amountAsset: String, priceAsset: String, timeStart: Date, period: String, dividers: [String] } -> [knex.QueryBuilder] */
const periodToQueries = ({
  amountAsset,
  priceAsset,
  timeStart,
  period,
  dividers,
}) => {
  const queries = [];

  let periodInterval = interval(period).getOrElse(null);

  if (!periodInterval) {
    return queries;
  }

  let itTimestamp = new Date(trunc(Unit.Minute, timeStart)).getTime();
  const endTimestamp = new Date(
    trunc(Unit.Minute, add(periodInterval, timeStart))
  ).getTime();

  while (itTimestamp < endTimestamp) {
    const theBiggestInterval = compose(
      find(i => div(periodInterval, i) >= 1),
      sortBy(
        compose(
          multiply(-1),
          prop('length')
        )
      ), // desc sorting
      map(d => interval(d).getOrElse(null))
    )(dividers);

    if (!theBiggestInterval) {
      break;
    }

    const timeEnd = add(theBiggestInterval, new Date(itTimestamp));

    queries.push(
      selectCandles({
        amountAsset,
        priceAsset,
        timeStart: new Date(itTimestamp),
        timeEnd: timeEnd,
        interval: theBiggestInterval.source,
      })
    );

    itTimestamp = timeEnd.getTime();
    periodInterval = intervalFromMillisecs(
      periodInterval.length - theBiggestInterval.length
    ).getOrElse(null);

    if (periodInterval === null) {
      break;
    }
  }

  return queries;
};

/** sql :: { String, String, { timeStart: Date, timeEnd: Date, interval: String } } -> String */
const sql = ({ amountAsset, priceAsset, params }) => {
  const paramsInterval = interval(params.interval).getOrElse(null);

  const ts = new Date(trunc(paramsInterval.unit, params.timeEnd));
  const te = new Date(trunc(Unit.Minute, params.timeEnd));

  const p = intervalFromMillisecs(te.getTime() - ts.getTime()).getOrElse('')
    .source;

  return pg('candles')
    .select(fieldsWithDecimals)
    .from(
      selectCandles({
        amountAsset,
        priceAsset,
        timeStart: params.timeStart,
        timeEnd: params.timeEnd,
        interval: params.interval,
      })
        .union(
          periodToQueries({
            amountAsset: amountAsset,
            priceAsset: priceAsset,
            timeStart: ts,
            period: p,
            dividers: ['1m', '5m', '15m', '30m', '1h', '3h', '6h', '12h', '1d'],
          })
        )
        .as('c')
    )
    .innerJoin(
      { a_dec: 'asset_decimals' },
      'c.amount_asset_id',
      'a_dec.asset_id'
    )
    .innerJoin(
      { p_dec: 'asset_decimals' },
      'c.price_asset_id',
      'p_dec.asset_id'
    )
    .orderBy('c.time_start', 'asc')
    .toString();
};

/** sql :: String -> String */
const assetDecimals = asset =>
  pg('asset_decimals')
    .select('decimals')
    .where('asset_id', asset)
    .toString();

module.exports = {
  highestDividerLessThen,
  assetDecimals,
  periodToQueries,
  sql,
};
