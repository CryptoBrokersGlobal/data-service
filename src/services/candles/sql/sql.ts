import * as knex from 'knex';
import { repeat } from 'ramda';
import { Interval, Unit, interval } from '../../../types';
import { add, trunc } from '../../../utils/date';
import {
  fromMilliseconds,
  unsafeIntervalsFromStrings,
  unsafeIntervalsFromStringsReversed,
} from '../../../utils/interval';
import { highestDividerLessThan, numberToUnitsPolynom } from './utils';

const pg = knex({ client: 'pg' });

const FIELDS = [
  'time_start',
  'amount_asset_id',
  'price_asset_id',
  'low',
  'high',
  'volume',
  'quote_volume',
  'max_height',
  'txs_count',
  'weighted_average_price',
  'open',
  'close',
  'interval_in_secs',
];

const FIELDS_WITH_DECIMALS: knex.ColumnName[] = [
  ...FIELDS,
  { a_dec: 'a_dec.decimals' },
  { p_dec: 'p_dec.decimals' },
];

const DIVIDERS = ['1m', '5m', '15m', '30m', '1h', '3h', '6h', '12h', '1d'];

export const selectCandles = ({
  amountAsset,
  priceAsset,
  timeStart,
  timeEnd,
  interval: inter,
}: {
  amountAsset: string;
  priceAsset: string;
  timeStart: Date;
  timeEnd: Date;
  interval: string;
}): knex.QueryBuilder =>
  pg('candles')
    .select(FIELDS)
    .where('amount_asset_id', amountAsset)
    .where('price_asset_id', priceAsset)
    .where('time_start', '>=', timeStart)
    .where('time_start', '<=', timeEnd)
    .where(
      'interval_in_secs',
      // should always be valid after validation
      highestDividerLessThan(
        interval(inter).unsafeGet(),
        unsafeIntervalsFromStrings(DIVIDERS)
      ).matchWith({
        Ok: ({ value: i }) => i.length / 1000,
        Error: ({ value: error }) => interval('1m').unsafeGet().length / 1000,
      })
    );

export const periodsToQueries = ({
  amountAsset,
  priceAsset,
  timeStart,
  periods,
}: {
  amountAsset: string;
  priceAsset: string;
  timeStart: Date;
  periods: Interval[];
}): knex.QueryBuilder[] => {
  const queries: knex.QueryBuilder[] = [];

  let itTimestamp = new Date(trunc(Unit.Minute, timeStart)).getTime();

  periods.forEach(period => {
    const timeEnd = add(period, new Date(itTimestamp));

    queries.push(
      selectCandles({
        amountAsset,
        priceAsset,
        timeStart: new Date(itTimestamp),
        timeEnd: timeEnd,
        interval: period.source,
      })
    );

    itTimestamp = timeEnd.getTime();
  });

  return queries;
};

export const sql = ({
  amountAsset,
  priceAsset,
  params: { timeStart, timeEnd, interval: inter },
}: {
  amountAsset: string;
  priceAsset: string;
  params: {
    timeStart: Date;
    timeEnd: Date;
    interval: string;
  };
}): string => {
  // should always be valid after validation
  const paramsInterval = interval(inter).unsafeGet();

  const ts = new Date(trunc(paramsInterval.unit, timeEnd));
  const te = new Date(trunc(Unit.Minute, timeEnd));
  const periodInMinutes = (te.getTime() - ts.getTime()) / (1000 * 60);

  const periodsForQueries = numberToUnitsPolynom(
    unsafeIntervalsFromStringsReversed(DIVIDERS).map(
      i => i.length / (1000 * 60)
    ),
    periodInMinutes
  ).reduce<Interval[]>((periods, polynom) => {
    return [
      ...periods,
      ...repeat(
        fromMilliseconds(polynom[0] * 1000 * 60).unsafeGet(),
        polynom[1]
      ),
    ];
  }, []);

  return pg('candles')
    .select(FIELDS_WITH_DECIMALS)
    .from(
      selectCandles({
        amountAsset,
        priceAsset,
        timeStart,
        timeEnd,
        interval: inter,
      })
        .union(
          periodsToQueries({
            amountAsset,
            priceAsset,
            timeStart: ts,
            periods: periodsForQueries,
          })
        )
        .as('c')
    )
    .innerJoin(
      { a_dec: 'asset_decimals' },
      'c.amount_asset_id',
      'a_dec.asset_id'
    )
    .innerJoin(
      { p_dec: 'asset_decimals' },
      'c.price_asset_id',
      'p_dec.asset_id'
    )
    .orderBy('c.time_start', 'asc')
    .toString();
};

export const assetDecimals = (asset: string): string =>
  pg('asset_decimals')
    .select('decimals')
    .where('asset_id', asset)
    .toString();

module.exports = {
  assetDecimals,
  periodsToQueries,
  sql,
};
