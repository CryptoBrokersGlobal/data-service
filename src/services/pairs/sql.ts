import * as knex from 'knex';
import { compose } from 'ramda';

const pg = knex({ client: 'pg' });

const COLUMNS = [
  'amount_asset_id',
  'price_asset_id',
  'first_price',
  'last_price',
  'volume',
  'quote_volume',
  'high',
  'low',
  'weighted_average_price',
  'txs_count',
  'volume_waves',
];

/** query :: Array[Object] -> String */
const query = (pairs: { amountAsset: string; priceAsset: string }[]) =>
  pg({ t: 'pairs' })
    .select(COLUMNS)
    .whereIn(
      ['t.amount_asset_id', 't.price_asset_id'],
      pairs.map(pair => [pair.amountAsset, pair.priceAsset])
    )
    .toString();

const searchAssets = (
  qb: knex.QueryBuilder,
  query: string,
  exactly: boolean,
  t: string = 't'
) =>
  qb
    .table({ [t]: 'assets' })
    .where(`${t}.asset_id`, query)
    .orWhere(`${t}.asset_name`, 'like', `${query}${exactly ? '' : '%'}`)
    .orWhere(`${t}.ticker`, 'like', `${query}${exactly ? '' : '%'}`)
    .orWhereRaw(
      `to_tsvector(${t}.asset_name) @@ to_tsquery('${query
        .split(' ')
        .join(' & ')}${exactly ? '' : ':*'}')`
    );

export const get = (pair: { amountAsset: string; priceAsset: string }) =>
  query([pair]);
export const mget = query;
export const search = ({
  search,
  limit,
}: {
  search: string;
  limit: number;
}) => {
  // asset - fuzzy search of amount or price assets
  // asset/ - exact search amount or price assets
  // /asset - same as asset/
  // asset1/asset2 - exact search of amount asset by asset1
  //                 and fuzzy search of price asset by asset2
  //                 or amount asset by asset2 and price asset by asset1
  const [asset1, priceAsset] = search.split('/');
  const amountAsset = !asset1.length ? priceAsset : asset1;

  const q = pg({ t: 'pairs' })
    .select(COLUMNS.map(column => `t.${column}`))
    .limit(limit);

  return compose(
    q => q.toString(),
    q => q.orderByRaw('volume_waves desc NULLS LAST'),
    (q: knex.QueryBuilder) =>
      priceAsset
        ? q
            .with('assets_cte', (qb: knex.QueryBuilder) =>
              searchAssets(qb, amountAsset, true, 'a')
                .crossJoin(qb =>
                  searchAssets(qb, priceAsset, false, 'p').as('cj')
                )
                .select([
                  { amount_asset_id: 'a.asset_id' },
                  { price_asset_id: 'cj.asset_id' },
                ])
            )
            .innerJoin(
              { direct_cte: 'assets_cte' },
              {
                'direct_cte.amount_asset_id': 't.amount_asset_id',
                'direct_cte.price_asset_id': 't.price_asset_id',
              }
            )
            .unionAll((qb: knex.QueryBuilder) =>
              qb
                .select(COLUMNS.map(column => `t1.${column}`))
                .from({ t1: 'pairs' })
                .innerJoin(
                  { reverse_cte: 'assets_cte' },
                  {
                    'reverse_cte.amount_asset_id': 't1.price_asset_id',
                    'reverse_cte.price_asset_id': 't1.amount_asset_id',
                  }
                )
            )
        : q
            .with('assets_cte', (qb: knex.QueryBuilder) =>
              searchAssets(qb, amountAsset, typeof priceAsset !== 'undefined')
            )
            .innerJoin(
              { amountCte: 'assets_cte' },
              'amountCte.asset_id',
              't.amount_asset_id'
            )
            .innerJoin(
              { priceCte: 'assets_cte' },
              'priceCte.asset_id',
              't.amount_asset_id'
            )
  )(q);
};
