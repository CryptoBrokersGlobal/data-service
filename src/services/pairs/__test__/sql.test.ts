import { get, mget, search } from '../sql';

describe('sql query from pairs', () => {
  it('should get one pair', () => {
    expect(
      get({
        amountAsset: '111',
        priceAsset: '222',
      })
    ).toMatchSnapshot();
  });

  it('should get many pairs', () => {
    expect(
      mget([
        {
          amountAsset: '111',
          priceAsset: '222',
        },
        {
          amountAsset: '333',
          priceAsset: '444',
        },
      ])
    ).toMatchSnapshot();
  });

  it('should search pairs for one asset', () => {
    expect(
      search({
        search: '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj',
        limit: 10,
      })
    ).toMatchSnapshot();
  });

  it('should search pairs for one asset exactly', () => {
    expect(
      search({
        search: '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj/',
        limit: 10,
      })
    ).toMatchSnapshot();
  });

  it('should search pairs for two assets (amount and price)', () => {
    expect(search({ search: 'BTC/WAVES', limit: 10 })).toMatchSnapshot();
  });
});
