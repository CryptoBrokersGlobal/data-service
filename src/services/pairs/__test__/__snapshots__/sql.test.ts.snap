// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sql query from pairs should get many pairs 1`] = `"select \\"amount_asset_id\\", \\"price_asset_id\\", \\"first_price\\", \\"last_price\\", \\"volume\\", \\"quote_volume\\", \\"high\\", \\"low\\", \\"weighted_average_price\\", \\"txs_count\\", \\"volume_waves\\" from \\"pairs\\" as \\"t\\" where (\\"t\\".\\"amount_asset_id\\", \\"t\\".\\"price_asset_id\\") in (('111', '222'), ('333', '444'))"`;

exports[`sql query from pairs should get one pair 1`] = `"select \\"amount_asset_id\\", \\"price_asset_id\\", \\"first_price\\", \\"last_price\\", \\"volume\\", \\"quote_volume\\", \\"high\\", \\"low\\", \\"weighted_average_price\\", \\"txs_count\\", \\"volume_waves\\" from \\"pairs\\" as \\"t\\" where (\\"t\\".\\"amount_asset_id\\", \\"t\\".\\"price_asset_id\\") in (('111', '222'))"`;

exports[`sql query from pairs should search pairs for one asset 1`] = `"with \\"assets_cte\\" as (select * from \\"assets\\" as \\"t\\" where \\"t\\".\\"asset_id\\" = '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj' or \\"t\\".\\"asset_name\\" like '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj%' or \\"t\\".\\"ticker\\" like '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj%' or to_tsvector(t.asset_name) @@ to_tsquery('7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj:*')) select \\"t\\".\\"amount_asset_id\\", \\"t\\".\\"price_asset_id\\", \\"t\\".\\"first_price\\", \\"t\\".\\"last_price\\", \\"t\\".\\"volume\\", \\"t\\".\\"quote_volume\\", \\"t\\".\\"high\\", \\"t\\".\\"low\\", \\"t\\".\\"weighted_average_price\\", \\"t\\".\\"txs_count\\", \\"t\\".\\"volume_waves\\" from \\"pairs\\" as \\"t\\" inner join \\"assets_cte\\" as \\"amountCte\\" on \\"amountCte\\".\\"asset_id\\" = \\"t\\".\\"amount_asset_id\\" inner join \\"assets_cte\\" as \\"priceCte\\" on \\"priceCte\\".\\"asset_id\\" = \\"t\\".\\"amount_asset_id\\" order by volume_waves desc NULLS LAST limit 10"`;

exports[`sql query from pairs should search pairs for one asset exactly 1`] = `"with \\"assets_cte\\" as (select * from \\"assets\\" as \\"t\\" where \\"t\\".\\"asset_id\\" = '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj' or \\"t\\".\\"asset_name\\" like '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj' or \\"t\\".\\"ticker\\" like '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj' or to_tsvector(t.asset_name) @@ to_tsquery('7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj')) select \\"t\\".\\"amount_asset_id\\", \\"t\\".\\"price_asset_id\\", \\"t\\".\\"first_price\\", \\"t\\".\\"last_price\\", \\"t\\".\\"volume\\", \\"t\\".\\"quote_volume\\", \\"t\\".\\"high\\", \\"t\\".\\"low\\", \\"t\\".\\"weighted_average_price\\", \\"t\\".\\"txs_count\\", \\"t\\".\\"volume_waves\\" from \\"pairs\\" as \\"t\\" inner join \\"assets_cte\\" as \\"amountCte\\" on \\"amountCte\\".\\"asset_id\\" = \\"t\\".\\"amount_asset_id\\" inner join \\"assets_cte\\" as \\"priceCte\\" on \\"priceCte\\".\\"asset_id\\" = \\"t\\".\\"amount_asset_id\\" order by volume_waves desc NULLS LAST limit 10"`;

exports[`sql query from pairs should search pairs for two assets (amount and price) 1`] = `"with \\"assets_cte\\" as (select \\"a\\".\\"asset_id\\" as \\"amount_asset_id\\", \\"cj\\".\\"asset_id\\" as \\"price_asset_id\\" from \\"assets\\" as \\"a\\" cross join (select * from \\"assets\\" as \\"p\\" where \\"p\\".\\"asset_id\\" = 'WAVES' or \\"p\\".\\"asset_name\\" like 'WAVES%' or \\"p\\".\\"ticker\\" like 'WAVES%' or to_tsvector(p.asset_name) @@ to_tsquery('WAVES:*')) as \\"cj\\" where \\"a\\".\\"asset_id\\" = 'BTC' or \\"a\\".\\"asset_name\\" like 'BTC' or \\"a\\".\\"ticker\\" like 'BTC' or to_tsvector(a.asset_name) @@ to_tsquery('BTC')) select \\"t\\".\\"amount_asset_id\\", \\"t\\".\\"price_asset_id\\", \\"t\\".\\"first_price\\", \\"t\\".\\"last_price\\", \\"t\\".\\"volume\\", \\"t\\".\\"quote_volume\\", \\"t\\".\\"high\\", \\"t\\".\\"low\\", \\"t\\".\\"weighted_average_price\\", \\"t\\".\\"txs_count\\", \\"t\\".\\"volume_waves\\" from \\"pairs\\" as \\"t\\" inner join \\"assets_cte\\" as \\"direct_cte\\" on \\"direct_cte\\".\\"amount_asset_id\\" = \\"t\\".\\"amount_asset_id\\" and \\"direct_cte\\".\\"price_asset_id\\" = \\"t\\".\\"price_asset_id\\" union all select \\"t1\\".\\"amount_asset_id\\", \\"t1\\".\\"price_asset_id\\", \\"t1\\".\\"first_price\\", \\"t1\\".\\"last_price\\", \\"t1\\".\\"volume\\", \\"t1\\".\\"quote_volume\\", \\"t1\\".\\"high\\", \\"t1\\".\\"low\\", \\"t1\\".\\"weighted_average_price\\", \\"t1\\".\\"txs_count\\", \\"t1\\".\\"volume_waves\\" from \\"pairs\\" as \\"t1\\" inner join \\"assets_cte\\" as \\"reverse_cte\\" on \\"reverse_cte\\".\\"amount_asset_id\\" = \\"t1\\".\\"price_asset_id\\" and \\"reverse_cte\\".\\"price_asset_id\\" = \\"t1\\".\\"amount_asset_id\\" order by volume_waves desc NULLS LAST limit 10"`;
